### ISSUE TITLE
[Airsonic] Scan (21)

### ISSUE THREAD

**Issue by tesshucom** [tesshucom/jpsonic#1736]
## Problem description

This issue was ported from the Airsonic archive. The top node for these issues is #1725.

It does not necessarily aim to solve all problems. However, some points of view will be fed back as Jpsonic is improved in the future.

It also includes issues that have already been resolved by Jpsonic at this stage. These will be progressively organized by topic.

## System information

 * **Jpsonic version**: since v111.5.4

## Note

 - Legacy code implementations are less testable. It is difficult to recover with minor corrections, and maintenance has almost reached its limit. Therefore, we are considering replacing it with another implementation that performs at least the same in v111.6.0.
   - Overall glitches will decrease and accuracy will increase. 
   - In addition, we aim to shift to a cyclical development that makes it easier to deal with any problems reported in the future.
 - No database design changes are planned. And the way users use it will largely remain the same.
   - Some feature options will be added.
   - If you're familiar with databases, you might be looking at transactional design. That's a very good point of view. However, that is not an immediate issue. If anything, it's the CRUD procedures that are most problematic. It is necessary to modify the implementation of the Service layer on the Java side first. We should always start with making it work right, verifying it, in that order.
     - It's not that I'm negative about improvements that focus on the DB layer. However, it should not be done in the same phase as the major overhaul of the main flow.
 -  As of v111.5.x, its scanning speed is faster than Subsonic even though it does Japanese processing. Since v111.6.x aims to achieve more accurate processing, the speed may be somewhat slower or faster, and there may be some effect.
    - However, parallelization will be supported as an option. It will probably be faster in environments where that feature is available.

## New task is issued

 - airsonic/airsonic#1708
   - -> #1840
  - **Clean up**
   - airsonic/airsonic#796
     - -> #1841
   - airsonic/airsonic#1425
   - airsonic/airsonic#1674
   - airsonic/airsonic#1756
     - -> #1842
 - **Request**
   - airsonic/airsonic#753
     - -> #1843
   - airsonic/airsonic#1046
     - -> #1844
 
## Closed

Tracking of the following agenda is closed.

 - airsonic/airsonic#359
   - Already added in Jpsonic
 - airsonic/airsonic#781
   - Since this is an irregular case, please use the forced scan option.
 - airsonic/airsonic#717
   - > These are the most useful I think.
     - Choosing which items are valid among the many tag standards is a difficult problem. 
     - Jpsonic defines specifications by the lowest common multiple with reference to SONY/Apple/Windows as a golden standard. By the way, compared to Subsonic/Airsonic, the only new additions are the Sort tag and Composer. These two items can also be effectively used from UPnP.
     - Some apps scale their implementations with their own sensitivities, while others determine a simple minimum configuration and maintain a realistically useful range. Jpsonic is the latter. Vendors that provide a variety of devices are familiar with generic user interfaces and know what items are commonly used by typical users. Managing more items than that is not necessarily an advantage of the software for many users.
 - airsonic/airsonic#732
   - Please check the timestamp. If a NAS is used, also the The time server of NAS.
   - Jpsonic has an option to force scan all and an option to force scan by album
 - airsonic/airsonic#772
   - Jpsonic has added an option to deal with this. Methods are available that do not rely on file timestamps.
   - However, it is unknown whether everything will go well like local disks and NAS. Keep in mind that using cloud libraries is more of a romantic endeavor than a widely accepted practice.
 - airsonic/airsonic#865
   - Opus is considered but Ape is not. Too much originality, and Windows is the main platform
 - airsonic/airsonic#1084
   - With Jpsonic you can do something similar. However, since it is a simple implementation for now, there is a possibility that it will be improved in the future. It should be more complicated than you think. Requires at least a redesign of the scan.
 - airsonic/airsonic#1095
   - Such fragmentary opinions are often expressed, but I think that the people who want to implement it and the people who want to use it are in the minority.
 - airsonic/airsonic#1189
   - It is thought that the cause is almost due to the tag format. 
     - Not all files edited by the tag editor are output in the correct format. Some apps may have problems.
     - Jpsonic uses a newer version of the tag parser than Airsonic. The error messages have also been slightly improved.
     - If you have any problems with Jpsonic, please issue a new Issues.
 - airsonic/airsonic#1441
   - Untraceable because the operating environment is not specified. If you reproduce it with Jpsonic, please report it again with the operating environment.
   - On the other hand, it is theoretically possible for this bug to occur in certain versions, and it is addressed in airsonic/airsonic#1550. It probably won't reproduce in Jpsonic.
 - airsonic/airsonic#1450
   - That is impossible. Not impossible, but terrifyingly complicated.
     - Probably, it seems that usage with high-speed access mode turned off in Subsonic is close to that. However, this has been removed in Jpsonic as certain procedures can cause overflow.
       - This is a testament to the complexity and risk involved in performing real-time view updates during a scan.
     - The scan is atomic from start to finish, and it is almost impossible to add anything else during the scan while maintaining integrity. Jpsonic in particular checks for missing tags, merges them, and creates sort keys. In other words, some operations are guaranteed to be consistent only after all libraly have been read. Processing flows require strong protection.
 - airsonic/airsonic#1800
   - If there is a cause, it is another matter. (usual timestamp issue) In Jpsonic, you may be able to take advantage of the force scan option and see what happens.
   - Since Jpsonic implements multi-genre, the implementation is partially different from other Sonic servers. Therefore, the on-memory genre master is referenced instead of the genre master on the DB. This cache will be cleared when scanning.
  - airsonic/airsonic#1671
    - It seems difficult to solve the problem with only the information you have provided. I might be able to parse the file if there is one...
      - If anything, I don't think iTunes is a great app for meta-handling. (If you consider versatility. If you only use Apple products, Apple's software may be useful.) Its specifications are somewhat special.
        -  See answers to `airsonic/airsonic#1189`.
        -  For Japanese, using the tag version converter can lead to data corruption. This means that there is an unresolved issue within iTunes as well.
      - Jpsonic has an updated parser and improved log messaging. Therefore, even when scanning the same file, the expected behavior may be different.
         -  See answers to `airsonic/airsonic#1477`.
       -  A relatively easy recovery method is to delete the iTunes tag and resave with Mp3tag or the like. 
  - airsonic/airsonic#1477
    - The latter mentioned is an issue that has nothing to do with the title of the agenda. Although the cause cannot be narrowed down with the presented information, there is a possibility of recovery by recreating the data folder.
    - Unlike Subsonic and Airsonic, Jpsonic has an implementation of messages for illegal files. So you'll see a one-line log instead of a stack trace. For example .... <table class="message">
<tr><td>WARN</td><td>Unable to read ***: [No audio header found within "***"]</td><td>When trying to read files other than music files, etc.</td></tr>
<tr><td>WARN</td><td>Unable to read ***: [Unable to read file because it is too small to be valid audio file: "***"]</td><td>When trying to read a file that cannot be treated as a music file due to file corruption, etc.</td></tr>
<tr><td>WARN</td><td>Unable to read cover art in ***: [Unable to read file because it is too small to be valid audio "***"]</td><td>When trying to load cover art from a corrupted file, etc.</td></tr>
<tr><td>WARN</td><td>Unable to read ***: ["***" Not a valid dsf file. Content does not start with 'DSD ']</td><td rowspan="5">If the tag format is invalid, if the extension is disguised and does not match the tag format, etc.</td></tr>
<tr><td>WARN</td><td>Unable to read ***: [Invalid Identification header for this Ogg File]</td></tr>
<tr><td>WARN</td><td>Unable to read ***: [OggS Header could not be found, not an ogg stream]</td></tr>
<tr><td>WARN</td><td>Unable to read ***: ["***" :java.lang.IllegalArgumentException: provided GUID is not supported by this reader.]</td></tr>
<tr><td>WARN</td><td>Unable to read ***: ["***" Wav RIFF Header not valid]</td></tr>
<tr><td>INFO</td><td>Only ID3 chunk is supported: ***</td><td>If you try to read a Wav that does not contain an ID3 chunk</td></tr>
</table>
